riscv/src/elf.rs

```diff
@@ -0,0 +1,1213 @@
+                rs1: Some(rs1),
+                rs2: None,
+            } => Ok((Register::new(*rd as u8), Register::new(*rs1 as u8))),
+            _ => Err(format!("Expected: rd, rs1, got {:?}", self.args)),
+        }
+    }
+
+    fn rrl(&self) -> Result<(Register, Register, String), Self::Error> {
+        match self.args {
+            HighLevelArgs {
+                imm: HighLevelImmediate::CodeLabel(addr),
+                rd: None,
+                rs1: Some(rs1),
+                rs2: Some(rs2),
+            } => Ok((
+                Register::new(*rs1 as u8),
+                Register::new(*rs2 as u8),
+                self.symbol_table.get_as_string(*addr),
+            )),
+            _ => Err(format!("Expected: rs1, rs2, label, got {:?}", self.args)),
+        }
+    }
+
+    fn rl(&self) -> Result<(Register, String), Self::Error> {
+        match self.args {
+            HighLevelArgs {
+                imm: HighLevelImmediate::CodeLabel(addr),
+                rd: None,
+                rs1: Some(rs1),
+                rs2: None,
+            } => Ok((
+                Register::new(*rs1 as u8),
+                self.symbol_table.get_as_string(*addr),
+            )),
+            HighLevelArgs {
+                imm: HighLevelImmediate::CodeLabel(addr),
+                rd: Some(rd),
+                rs1: None,
+                rs2: None,
+            } => Ok((
+                Register::new(*rd as u8),
+                self.symbol_table.get_as_string(*addr),
+            )),
+            _ => Err(format!("Expected: {{rs1|rd}}, label, got {:?}", self.args)),
+        }
+    }
+
+    fn rro(&self) -> Result<(Register, Register, u32), Self::Error> {
+        match self.args {
+            HighLevelArgs {
+                imm: HighLevelImmediate::Value(imm),
+                rd: Some(rd),
+                rs1: Some(rs1),
+                rs2: None,
+            } => Ok((
+                Register::new(*rd as u8),
+                Register::new(*rs1 as u8),
+                *imm as u32,
+            )),
+            HighLevelArgs {
+                imm: HighLevelImmediate::Value(imm),
+                rd: None,
+                rs1: Some(rs1),
+                rs2: Some(rs2),
+            } => Ok((
+                Register::new(*rs2 as u8),
+                Register::new(*rs1 as u8),
+                *imm as u32,
+            )),
+            _ => Err(format!(
+                "Expected: {{rd, rs1 | rs2, rs1}}, imm, got {:?}",
+                self.args
+            )),
+        }
+    }
+
+    fn empty(&self) -> Result<(), Self::Error> {
+        match self.args {
+            HighLevelArgs {
+                imm: HighLevelImmediate::None,
+                rd: None,
+                rs1: None,
+                rs2: None,
+            } => Ok(()),
+            _ => Err(format!("Expected: no args, got {:?}", self.args)),
+        }
+    }
+}
+
+struct AddressMap<'a>(BTreeMap<u32, &'a program_header::ProgramHeader>);
```

Comment:
Is the key an address? What is the value?

Reply:
Added comments.

Reply:
Address to program section.